File: Baekjoon/Problem_2000~2999/Problem_2225/Problem_2225.cpp
#include <cstdio>
#include <algorithm>

using namespace std;
const int MOD = 1000000000;
const int MAXN = 200;

int dp[MAXN+1][MAXN+1];

int solve(const int n, const int k) {
    // dp base
    if (dp[n][k] >= 0) return dp[n][k];

    // recursion base
    else if (k == 0) {
        dp[n][k] = (n == 0);
        return dp[n][k];
    }

    // recursion default
    dp[n][k] = 0;
    for (int i = 0; i <= n; i++) {
        dp[n][k] = (dp[n][k] + solve(n-i, k-1)) % MOD;
    }
    return dp[n][k];
}

int main() {
    int N, K; scanf("%d %d", &N, &K);

    for (int r = 0; r <= N; r++) {
        fill(dp[r], dp[r]+K+1, -1);
    }

    printf("%d\n", solve(N, K));

    return 0;
}

File: Baekjoon/Problem_2000~2999/Problem_2225/Problem_2225.py
import sys

input = lambda: sys.stdin.readline().rstrip()

MOD = 1_000_000_000

def solve(n, k):
    # dp base
    if dp[n][k] >= 0:
        return dp[n][k]
    
    # recursion base
    if k == 0:
        dp[n][k] = (n == 0)
        return dp[n][k]
    
    # recursion default
    dp[n][k] = 0
    for i in range(n+1):
        dp[n][k] = (dp[n][k] + solve(n-i, k-1)) % MOD
    return dp[n][k]

if __name__ == "__main__":
    N, K = map(int, input().split())
    
    dp = [ [ -1 ] * (K+1) for _ in range(N+1) ]
    
    print(solve(N, K))
            